Update SSO docs for Azure AD (#2070)
TykTechnologies/tyk-docs: tyk-docs/content/tyk-stack/tyk-manager/sso/dashboard-login-azure-sso.md

  - /advanced-configuration/integrate/sso/dashboard-login-azure-sso/
---

This is an end-to-end worked example of how you can use [AzureAD](https://azure.microsoft.com/en-gb/services/active-directory/) and our [Tyk Identity Broker (TIB)](https://tyk.io/docs/concepts/tyk-components/identity-broker/
) to log in to your Dashboard.
This guide assumes the following:

> * You already have authorised access to Tyk's Dashboard. If you haven't, [get the authorisation key by following this doc](/docs/basic-config-and-security/security/dashboard/create-users/#create-a-dashboard-user-with-the-api).

Reviewer: ```suggestion
* You already have authorised access to Tyk's Dashboard. If you haven't, [get the authorisation key by following this doc]({{< ref "basic-config-and-security/security/dashboard/create-users/#create-a-dashboard-user-with-the-api">}}).
```

> * You already have authorised access to Tyk's Dashboard. If you haven't, [get the authorisation key by following this doc]({{< ref "/basic-config-and-security/security/dashboard/create-users/#create-a-dashboard-user-with-the-api">}}).

Reviewer 2: ```suggestion
* You already have authorised access to Tyk's Dashboard. If you haven't, [get the authorisation key by following this doc]({{< ref "basic-config-and-security/security/dashboard/create-users#create-a-dashboard-user-with-the-api">}}).
```

> * You already have authorised access to Tyk's Dashboard. If you haven't, [get the authorisation key by following this doc](/docs/basic-config-and-security/security/dashboard/create-users/#a-name-with-api-a-create-a-dashboard-user-with-the-api).

Reviewer: ```suggestion
* You already have authorised access to Tyk's Dashboard. If you haven't, [get the authorisation key by following this doc](/docs/basic-config-and-security/security/dashboard/create-users/#create-a-dashboard-user-with-the-api).
```

## Azures's side
1. Access your Azure Portal and navigate to the Azure Active Directory page.
2. Go to app registrations and create or access an application you want to use for Dashboard access.
  - If you are creating an application, give it a name and register it 
3. Add a redirect URL to your application as callback to TIB in your Azure application:
  - In your app, either via the Authentication menu or the redirect URL shortcut navigate to and add the redirect to TIB in the Web category i.e. `http://localhost:3000/auth/{PROFILE-NAME-IN-TIB}/openid-connect/callback`.

>     ![redirect-url](/docs/img/azureAD/redirect-URL-1.png)

Reviewer: ```suggestion
    ![redirect-url]({{< ref "img/azureAD/redirect-URL-1.png" >}})
```

>     ![redirect-url]({{< ref "img/azureAD/redirect-URL-1.png" >}})

Reviewer: ```suggestion
    ![redirect-url]({{< ref "/img/azureAD/redirect-URL-1.png" >}})
```

> 4. Go to Overview and add a secret in Client Credentials. Don't forget to copy the secret value- not the secretID. 

Reviewer: ```suggestion
4. Go to Overview and add a secret in Client Credentials. Don't forget to copy the secret value, not the secretID. 
```

>     ![overview]({{< ref "img/azureAD/overview-1.png">}})

Reviewer: ```suggestion
    ![overview]({{< ref "/img/azureAD/overview-1.png">}})
```

Check Microsoft's [documentation](https://docs.microsoft.com/en-us/azure/active-directory/develop/quickstart-register-app) for more detail.

## Dashboard's side 
1. Log in to your dashboard and select Identity Management, located under System Management
2. Create a profile and select OpenID Connect as the provider type

> 3. Under Profile Configuration, paste the secret value, clientID, and Discover URL from the Azure site. 

Reviewer: ```suggestion
3. Under Profile Configuration, paste the secret value, clientID, and well-known endpoint URL from the Azure site. 
```

  - Profile Configuation may look something like this:

  ![profile-configuration](/docs/img/azureAD/profile-configuration-1.png)

>   - The Discover URL is created by Azure and can be located by selecting Endpoints on their site

Reviewer: ```suggestion
  - The well-known endpoint URL is created by Azure and can be located by selecting Endpoints on their site
```

>   ![endpoints]({{< ref "img/azureAD/endpoints-11.png" >}})

Reviewer: ```suggestion
  ![endpoints]({{< ref "/img/azureAD/endpoints-11.png" >}})
```

> 8. Test that it works:

Reviewer: ```suggestion
4. Test that it works:
```

> 4. Test that it works:

Author: images and header added

   From the browser call `http://localhost:3000/auth/{PROFILE-NAME-IN-TIB}/openid-connect`
    - If it's working you'll be redirected to Azures's web page and will be asked to enter your Azure user name and password.
    - If you were successfully authenticated by Azure then you'll be redirected to the Tyk Dashboard and login into it without going through the login page. Job's done!

## Enhancements

Once it's working you can also add more enhancements such as automatic user group mapping from your AzureAD security groups or users groups to Tyk Dashboards groups.

### User group mapping

> Group mapping can be managed in Advanced Settings by passing an identity provider and selecting a Tyk User Group which can be created in the the User Groups section of

Reviewer 3: _Maybe consider editing to something like below_

Group mapping can be managed from Advanced Settings section of the Profile Configuration screen. 

(show Profile Configuration screen here)

As illustrated in the screen below the following information must be provided:
- Identity provider role
- Tyk User Group: This can be created from the User Groups section of the dashboard (reference a link to a page in tyk docs here to show how to create a user group). When creating your User Group, one can also select and adjust the permissions for each group. (add a link to existing docs that shows how to select and adjust permissions for a user group)

(include the screen that shows mapping from Azure identity to Tyk User group mapping here)

You can select the scopes you would like your request to include. By default, Tyk will provide the _connectid_ scope, anything additional must be requested.

<Maybe add a description of the other optional user group mapping fields here, e.g. Fallback user group ad custom group name>

the dashboard. When creating your User Group, one can also select and adjust the permissions for each group.

In the Advanced Settings of the Provider Configuration, you can select the scopes you would like your request to include. By default, Tyk will provide the connectid scope, anything additional must be requested. 

> ![profile-configuration-additional-options]({{< ref "img/azureAD/additional-options.png">}})

Reviewer: ```suggestion
![profile-configuration-additional-options]({{< ref "/img/azureAD/additional-options.png">}})
```

> ![profile-configuration-raw-editor]({{< ref "img/azureAD/raw-editor.png">}})

Reviewer: ```suggestion
![profile-configuration-raw-editor]({{< ref "/img/azureAD/raw-editor.png">}})
```

> For debugging purposes, you can find an example we created using the OpenID Connect playground.

Reviewer 3: Make this a new section, titled OpenID Connect example

> 1. Add the redirect url found on the OpenID Connect site to the redirect urls found under the Web section

Reviewer 3: Show a screenshot of how to access the web section here, is it included somewhere on dashboard?

Author: added

> ![additional-url-added]({{< ref "img/azureAD/openid_connect/additional_redirect_url.png">}})

Reviewer: ```suggestion
![additional-url-added]({{< ref "img/azureAD/openid_connect/additional_redirect_url.png">}})
```

> ![additional-url-added]({{< ref "img/azureAD/openid_connect/additional_redirect_url.png">}})

Reviewer: ```suggestion
![additional-url-added]({{< ref "/img/azureAD/openid_connect/additional_redirect_url.png">}})
```

> 2. Copy the OpenID Connect endpoint

Reviewer 3: Accompany steps with additional screenshots, e.g. include a screenshot of the screen the user clicks edit on for the OpenID connect site where the Server Template dropdown exists.

Which screen do copy and paste the Client ID and Client Secret into?
Which screen do press Start on

Author: added

> 2. Copy the OpenID Connect endpoint

Reviewer 3: Add a space before step 2. When previewing in the browser all the number steps are displayed in one paragraph

3. On the OpenID Connect site select Edit. In the Server Template dropdown menu select the Custom option and paste the endpoint in the Discovery Document URL. 
4. Press the Use Discovery Document button and this will autofill Authorization Token Endpoint, Token Endpoint, and Token Keys Endpoint
5. Copy and paste the Client ID and Client Secret. Scope is autofilled for you and save the configuration.

> ![openid-connect-step-1]({{< ref "img/azureAD/openid_connect/step-1.png">}})

Reviewer: ```suggestion
![openid-connect-step-1]({{< ref "/img/azureAD/openid_connect/step-1.png">}})
```

6. Press start and if done correctly, this should prompt you to sign in to your Azure account.

> ![openid-connect-step-2]({{< ref "img/azureAD/openid_connect/step-2.png">}})

Reviewer: ```suggestion
![openid-connect-step-2]({{< ref "/img/azureAD/openid_connect/step-2.png">}})
```

7. You should then be redirected back to OpenID Connect where you'll be shown the Exchange Code. This needs to be turned into an access token. Press the exchange button under the request and then press Next.

> ![openid-connect-step-3]({{< ref "img/azureAD/openid_connect/step-3.png">}})

Reviewer: ```suggestion
![openid-connect-step-3]({{< ref "/img/azureAD/openid_connect/step-3.png">}})
```

> ![openid-connect-step-4]({{< ref "img/azureAD/openid_connect/step-4.png">}})

Reviewer: ```suggestion
![openid-connect-step-4]({{< ref "/img/azureAD/openid_connect/step-4.png">}})
```

8. We can then verify this by pressing the verify button. We can also view the information or scope of what is being returned by heading to jwt.io and viewing the payload: data there.

> ![openid-connect-step-5](/docs/img/azureAD/openid_connect/step-5.png)

Reviewer: ```suggestion
![openid-connect-step-5]({{< ref "img/azureAD/openid_connect/step-5.png">}})
```

> ![openid-connect-step-5]({{< ref "img/azureAD/openid_connect/step-5.png">}})

Reviewer: ```suggestion
![openid-connect-step-5]({{< ref "/img/azureAD/openid_connect/step-5.png">}})
```

9. We are given an object with key, value pairs and we can pass in the key ie. name to our Custom User Group and the value of to our Identity Provider Role in our Tyk dashboard as shown in the example above. 

> ![openid-connect-step-6](/docs/img/azureAD/openid_connect/step-6.png)

Reviewer: ```suggestion
![openid-connect-step-6]({{< ref "img/azureAD/openid_connect/step-6.png">}})
```

> ![openid-connect-step-6]({{< ref "img/azureAD/openid_connect/step-6.png">}})

Reviewer: ```suggestion
![openid-connect-step-6]({{< ref "/img/azureAD/openid_connect/step-6.png">}})
```

This guide assumes the following:

* You already have authorised access to Tyk's Dashboard. If you haven't, [get the authorisation key by following this doc](/docs/basic-config-and-security/security/dashboard/create-users/#a-name-with-api-a-create-a-dashboard-user-with-the-api).

## Azures's side
1. Access your Azure Portal and navigate to the Azure Active Directory page.
2. Go to app registrations and create or access an application you want to use for Dashboard access.
  - if you are creating an application, give it a name and register it 
3. Add a redirect URL to your application as callback to TIB in your Azure application:
  - In your app, either via the Authentication menu or the redirect URL shortcut navigate to and add the redirect to TIB in the Web category i.e. `http://localhost:3000/auth/{PROFILE-NAME-IN-TIB}/openid-connect/callback`.
    ![redirect-url](/docs/img/azureAD/redirect-URL.png)
4. Go to Overview and add a secret in Client Credentials. Don't forget to copy the secret value- not the secretID. 
    ![overview](/docs/img/azureAD/overview.png)

> Check [Microsoft's documentation](https://docs.microsoft.com/en-us/azure/active-directory/develop/quickstart-register-app) for more detail.

Reviewer 4: I think this apostrophe in Microsoft's needs to be escaped since it's rendering like:

Microsoft&rsquo;s

Reviewer 4: Rendering like: Microsoft& rsquo;s

Author: done

  ![endpoints] (/docs/img/azureAD/endpoints.png)

8. Test that it works:
   From the browser call `http://localhost:3000/auth/{PROFILE-NAME-IN-TIB}/openid-connect`
    - If it's working you'll be redirected to Azures's web page and will be asked to enter your Azure user name and password.
    - If you were successfully authenticated by Azure then you'll be redirected to the Tyk Dashboard and login into it without going through the login page. Job's done!

## Enhancements

Once it's working you can also add more enhancements such as automatic user group mapping from your AzureAD security groups or users groups to Tyk Dashboards groups.

### User group mapping
Group mapping can be managed in Advanced Settings by passing an identity provider and selecting a Tyk User Group which can be created in the the User Groups section of
the dashboard. When creating your User Group, one can also select and adjust the permissions for each group.

In the Advanced Settings of the Provider Configuration, you can select the scopes you would like your request to include. By default, Tyk will provide the connectid scope, anything additional must be requested. 

> ![profile-configuration-additional-options] (/docs/img/azureAD/additional_options.png)

Reviewer 4: I believe the space is preventing the image from rendering correctly
i.e. it says
![profile-configuration-additional-options] (/docs/img/azureAD/additional_options.png)

When it should be
![profile-configuration-additional-options](/docs/img/azureAD/additional_options.png)

Author: I removed the spacing here as well

![profile-configuration-raw-editor](/docs/img/azureAD/raw_editor.png)

For debugging purposes, you can find an example we created using the OpenID Connect playground.
1. Add the redirect url found on the OpenID Connect site to the redirect urls found under the Web section

> ![additional-url-added](/docs/img/azureAD/openid_connect/additional_redirect_url_added)

Reviewer 4: I think you're missing an extension here, the image is broken
instead of 
![additional-url-added](/docs/img/azureAD/openid_connect/additional_redirect_url_added)

try

![additional-url-added](/docs/img/azureAD/openid_connect/additional_redirect_url_added.png)

Reviewer 4: i.e. add a .png after your filename additional_redirect_url_added to make it "additional_redirect_url_added.png"

Author: good catch! thank you and done

> ![additional-url-added] (/docs/img/azureAD/openid_connect/additional_redirect_url_added)

Reviewer 4: Potentially missing extension to the file here?

Author: I removed the spacing here

This guide assumes the following:

* You already have authorised access to Tyk's Dashboard. If you haven't, [get the authorisation key by following this doc](/docs/basic-config-and-security/security/dashboard/create-users/#a-name-with-api-a-create-a-dashboard-user-with-the-api).

## Azures's side
1. Access your Azure Portal and navigate to the Azure Active Directory page.
2. Go to app registrations and create or access an application you want to use for Dashboard access.
  - if you are creating an application, give it a name and register it 
3. Add a redirect URL to your application as callback to TIB in your Azure application:
  - In your app, either via the Authentication menu or the redirect URL shortcut navigate to and add the redirect to TIB in the Web category i.e. `http://localhost:3000/auth/{PROFILE-NAME-IN-TIB}/openid-connect/callback`.
    (/docs/img/azureAD/redirect-URL.png)
4. Go to Overview and add a secret in Client Credentials. Don't forget to copy the secret value- not the secretID. 

>     (/docs/img/azureAD/overview.png)

Reviewer 2: Please fix this so the image is added correctly throughout the page

Reviewer: @Monica-Cedeno you need to add the square brackets with the description of the image before each image :)

Author: square brackets and spacing was resolved

> See their documentation for more detail: https://docs.microsoft.com/en-us/azure/active-directory/develop/quickstart-register-app

Reviewer 2: ```suggestion
Check [Microsoft's documentation](https://docs.microsoft.com/en-us/azure/active-directory/develop/quickstart-register-app) for more detail.
```
